src/ray/gcs/tables.h

```diff
@@ -337,26 +360,105 @@ class Table : private Log<ID, Data>,
   using Log<ID, Data>::command_type_;
   using Log<ID, Data>::GetRedisContext;
 
- private:
   int64_t num_adds_ = 0;
   int64_t num_lookups_ = 0;
 };
 
-class ObjectTable : public Log<ObjectID, ObjectTableData> {
+template <typename ID, typename Data>
+class SetInterface {
+ public:
+  using DataT = typename Data::NativeTableType;
+  using WriteCallback = typename Log<ID, Data>::WriteCallback;
+  virtual Status Add(const JobID &job_id, const ID &id, std::shared_ptr<DataT> &data,
+                     const WriteCallback &done) = 0;
+  virtual Status Remove(const JobID &job_id, const ID &id, std::shared_ptr<DataT> &data,
+                        const WriteCallback &done) = 0;
+  virtual ~SetInterface(){};
+};
+
+/// \class Set
+///
+/// A GCS table where every entry is an addable & removable set. This class is not
+/// meant to be used directly. All set classes should derive from this class
+/// and override the prefix_ member with a unique prefix for that set, and the
+/// pubsub_channel_ member if pubsub is required.
+///
+/// Example tables backed by Set:
+///   ObjectTable: Stores a set of which clients have added an object.
+template <typename ID, typename Data>
+class Set : private Log<ID, Data>,
+            public SetInterface<ID, Data>,
+            virtual public PubsubInterface<ID> {
```

Comment:
It looks a bit weird to me that `Set` inherits `Log`, so is `Table`.  It feels more natural to have an abstract super class, say `Collection`, and have `Set`, `Log` and `Table` inherit it.

But It's fine to do that later. 

@stephanie-wang any thoughts?

Reply:
Yes, that sounds like a good idea! Agreed that we should do it in a future PR.